src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -18,29 +29,217 @@ static UCS_CLASS_INIT_FUNC(uct_cuda_ipc_ep_t, uct_iface_t *tl_iface,
+            if (cu_ret != CUDA_SUCCESS) {                               \
+                cuGetErrorString(cu_ret, &cu_err_str);                  \
+                ucs_error("cuPointerGetAttribute failed ret:%s", cu_err_str); \
+                return UCS_ERR_IO_ERROR;                                \
+            }                                                           \
+        }                                                               \
+        if ((CUDA_SUCCESS == cu_err) && (key->dev_num == (int) cu_device) && \
+            (local_ptr_ctx == remote_ptr_ctx)) {                        \
+            mapped_rem_addr = (void *) remote_addr;                     \
+        }                                                               \
+        else {                                                          \
+            /* Is uintptr_t equivalent to uint64_t?  */                 \
+            mapped_rem_base_addr = uct_cuda_ipc_ep_attach_rem_seg(ep, iface, \
+                                                                  key); \
+            offset = (uintptr_t) remote_addr - (uintptr_t) key->d_rem_bptr; \
+            if (offset > key->b_rem_len) {                              \
+                ucs_fatal("Access memory outside memory range attempt\n"); \
+            }                                                           \
+            mapped_rem_addr = (void *) ((uintptr_t) mapped_rem_base_addr + offset); \
+        }                                                               \
+    } while(0);
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_ipc_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
+                                  size_t length, uct_cuda_ipc_iface_t *iface,
+                                  uct_cuda_ipc_key_t *key, uct_completion_t *comp)
 {
-    return UCS_OK;
+    ucs_status_t status = UCS_OK;
+    uct_cuda_ipc_event_desc_t *cuda_ipc_event;
+    CUstream stream;
+    ucs_queue_head_t *outstanding_queue;
+
+    if (!length) return UCS_OK;
+    if (0 == iface->streams_initialized) {
+        status = uct_cuda_ipc_iface_init_streams(iface);
+        if (UCS_OK != status) return status;
+    }
+    stream = iface->stream_d2d[key->dev_num];
+    outstanding_queue = &iface->outstanding_d2d_event_q;
+    cuda_ipc_event = ucs_mpool_get(&iface->event_desc);
+    if (ucs_unlikely(cuda_ipc_event == NULL)) {
+        ucs_error("Failed to allocate cuda_ipc event object");
+        return UCS_ERR_NO_MEMORY;
+    }
+    status = UCT_CUDADRV_FUNC(cuMemcpyDtoDAsync((CUdeviceptr) dst,
+                                                (CUdeviceptr) src,
+                                                length, stream));
+    if (UCS_OK != status) {
+        ucs_error("cuMemcpyDtoD Failed");
+        return UCS_ERR_IO_ERROR;
+    }
+    status = UCT_CUDADRV_FUNC(cuEventRecord(cuda_ipc_event->event, stream));
+    if (UCS_OK != status) {
+        ucs_error("cuEventRecord Failed");
```

Comment:
no need of ucs_error again. it already printed in UCT_CUDADRV_FUN.   we can remot it at all the places

Reply:
Handled this.